From e6df42b8734cb01aaefbc6b734fc93de5970bbc5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lu=C3=ADs=20Pereira?= <luis.artur.pereira@gmail.com>
Date: Mon, 14 Jul 2014 13:39:50 -0700
Subject: [PATCH] Qt5: Use XDG_DATA_DIR for themeHint::IconThemeSearchPaths

Copied from:
    qtbase/src/platformsupport/themes/genericunix/qgenericunixthemes.cpp

and made it an const member.
---
 qt5/lxqtplatformtheme.cpp | 27 ++++++++++++++++++++++++---
 qt5/lxqtplatformtheme.h   |  2 ++
 2 files changed, 26 insertions(+), 3 deletions(-)

diff --git a/qt5/lxqtplatformtheme.cpp b/qt5/lxqtplatformtheme.cpp
index f6df924..88a0453 100644
--- a/qt5/lxqtplatformtheme.cpp
+++ b/qt5/lxqtplatformtheme.cpp
@@ -40,6 +40,8 @@
 #include <QToolBar>
 #include <QSettings>
 #include <QTimer>
+#include <QDir>
+#include <QFileInfo>
 #include <QFileSystemWatcher>
 #include "qiconloader_p.h"
 
@@ -233,9 +235,8 @@ QVariant LXQtPlatformTheme::themeHint(ThemeHint hint) const {
         return iconTheme_;
     case SystemIconFallbackThemeName:
         return "hicolor";
-    case IconThemeSearchPaths: // FIXME: should use XDG_DATA_DIRS instead
-        return QStandardPaths::locateAll(QStandardPaths::GenericDataLocation, QStringLiteral("icons"), QStandardPaths::LocateDirectory)
-            << QStandardPaths::locate(QStandardPaths::HomeLocation, QStringLiteral(".icons"), QStandardPaths::LocateDirectory);
+    case IconThemeSearchPaths:
+        return xdgIconThemePaths();
     case StyleNames:
         // qDebug() << "StyleNames";
         return QStringList() << style_;
@@ -275,3 +276,23 @@ QVariant LXQtPlatformTheme::themeHint(ThemeHint hint) const {
     }
     return QPlatformTheme::themeHint(hint);
 }
+
+// Helper to return the icon theme paths from XDG.
+QStringList LXQtPlatformTheme::xdgIconThemePaths() const
+{
+    QStringList paths;
+    // Add home directory first in search path
+    const QFileInfo homeIconDir(QDir::homePath() + QStringLiteral("/.icons"));
+    if (homeIconDir.isDir())
+        paths.prepend(homeIconDir.absoluteFilePath());
+
+    QString xdgDirString = QFile::decodeName(qgetenv("XDG_DATA_DIRS"));
+    if (xdgDirString.isEmpty())
+        xdgDirString = QLatin1String("/usr/local/share/:/usr/share/");
+    foreach (const QString &xdgDir, xdgDirString.split(QLatin1Char(':'))) {
+        const QFileInfo xdgIconsDir(xdgDir + QStringLiteral("/icons"));
+        if (xdgIconsDir.isDir())
+            paths.append(xdgIconsDir.absoluteFilePath());
+    }
+    return paths;
+}
diff --git a/qt5/lxqtplatformtheme.h b/qt5/lxqtplatformtheme.h
index c8e04af..abba806 100644
--- a/qt5/lxqtplatformtheme.h
+++ b/qt5/lxqtplatformtheme.h
@@ -91,6 +91,8 @@ private Q_SLOTS:
     QVariant cursorFlashTime_;
     QFileSystemWatcher *settingsWatcher_;
     QString settingsFile_;
+
+    QStringList xdgIconThemePaths() const;
 };
 
 #endif // LXQTPLATFORMTHEME_H
